File: platforms/android/module-api/src/main/kotlin/rs/whisker/runtime/ModuleDefinition.kt
// Phase L-2a — `ModuleDefinition` DSL surface (Android).
//
// Kotlin counterpart of `platforms/ios/Sources/WhiskerModuleApi/
// ModuleDefinition.swift`. Modeled after Expo Modules'
// `ModuleDefinition` (https://docs.expo.dev/modules/module-api/).
//
// ## Target syntax
//
// ```kotlin
// class VideoModule : WhiskerModule() {
//     override fun definition() = ModuleDefinition {
//         Name("Video")
//
//         Constants("maxResolution" to "1080p")
//
//         View(WhiskerVideoComponent::class.java) {
//             Prop("src") { view: WhiskerVideoComponent, value: String -> view.setSrc(value) }
//             Function("play")  { view: WhiskerVideoComponent -> view.play()  }
//             Function("pause") { view: WhiskerVideoComponent -> view.pause() }
//             Function("seek")  { view: WhiskerVideoComponent, seconds: Double -> view.seek(seconds) }
//             Events("onCompleted")
//         }
//     }
// }
// ```
//
// Function-only modules omit the inner `View(...)` block:
//
// ```kotlin
// class LocalStoreModule : WhiskerModule() {
//     override fun definition() = ModuleDefinition {
//         Name("WhiskerLocalStore")
//         Function("save") { key: String, value: String ->
//             prefs.edit().putString(key, value).apply()
//             true
//         }
//         Function("load") { key: String -> prefs.getString(key, null) }
//     }
// }
// ```
//
// ## What L-2a delivers
//
// Type surface only. The `WhiskerModule` base + DSL types compile
// and authors can write modules using the syntax above; the KSP
// codegen that turns `definition()` into Lynx prop / method
// registrations lands in Phase L-2c. Existing `@WhiskerComponent`
// annotation surface continues in parallel.

package rs.whisker.runtime

// ----- Component model -----------------------------------------------------

/**
 * Type-erased component the DSL collects. Concrete subtypes live
 * below. Authors normally don't reference [WhiskerDefinitionComponent]
 * directly — the factory functions (`Name`, `Prop`, `Function`,
 * etc.) return the right subtype.
 */
public sealed interface WhiskerDefinitionComponent

/** `Name("Foo")` — the module's local tag name. */
public data class WhiskerNameComponent(public val value: String) :
    WhiskerDefinitionComponent

/**
 * `Constants("k" to v, ...)` — static key/value pairs exposed to
 * the host. Dictionary form only in v1; the dynamic closure form
 * and per-key lazy `Constant("k") { ... }` form land later.
 */
public data class WhiskerConstantsComponent(public val values: Map<String, Any?>) :
    WhiskerDefinitionComponent

/**
 * `View(Foo::class.java) { ... }` — registers a Lynx UI subclass
 * + its inner DSL block (Prop / Function / Events). The class is
 * type-erased to `Class<*>` so the parent struct isn't generic;
 * the concrete class is the Lynx UI subclass (typically a
 * [WhiskerUI] subclass).
 */
public data class WhiskerViewComponent(
    public val viewClass: Class<*>,
    public val components: List<WhiskerDefinitionComponent>,
) : WhiskerDefinitionComponent

/**
 * Type-erased prop setter the framework calls on prop dispatch.
 * `view` is the Lynx UI instance; `value` is the raw decoded
 * Lynx value (typed-decoded by L-2c at dispatch time).
 */
public typealias WhiskerPropSetterFn = (view: Any, value: Any?) -> Unit

public data class WhiskerPropComponent(
    public val name: String,
    public val setter: WhiskerPropSetterFn,
) : WhiskerDefinitionComponent

/**
 * Type-erased function handler. `view` is `null` for
 * module-level [Function]s, the Lynx UI instance for view-block
 * [Function]s. `args` carry positional arguments from the JS /
 * Rust call site; the return value is auto-encoded by the
 * dispatch glue (`Unit` → `WhiskerValue.Null`).
 */
public typealias WhiskerFunctionHandlerFn = (view: Any?, args: List<Any?>) -> Any?

public data class WhiskerFunctionComponent(
    public val name: String,
    public val handler: WhiskerFunctionHandlerFn,
) : WhiskerDefinitionComponent

/**
 * `Events("a", "b", ...)` — declare event names this module
 * emits. Just metadata; dispatch stays imperative via
 * `WhiskerCustomEvent.dispatch(...)`.
 */
public data class WhiskerEventsComponent(public val names: List<String>) :
    WhiskerDefinitionComponent

// ----- DSL builders --------------------------------------------------------

/**
 * Top-level builder. Authors call DSL factory functions inside
 * the lambda passed to [ModuleDefinition]; the builder collects
 * the resulting components.
 *
 * Marked `@DslMarker` so the inner [WhiskerViewDefinitionBuilder]
 * doesn't expose top-level factories like [View] that would be
 * nonsensical inside a `View(...) { ... }` block.
 */
@DslMarker
public annotation class WhiskerDefinitionDsl

@WhiskerDefinitionDsl
public class WhiskerModuleDefinitionBuilder {
    /**
     * Internal mutable collection of components the factory
     * functions ([Name], [View], [Function], [Constants], [Events])
     * append to as they execute. The lambda block's receiver is a
     * [WhiskerModuleDefinitionBuilder], so calls inside the block
     * resolve to the extension functions defined below — each one
     * has the side effect of appending its component here.
     */
    @PublishedApi
    internal val components: MutableList<WhiskerDefinitionComponent> = mutableListOf()
}

@WhiskerDefinitionDsl
public class WhiskerViewDefinitionBuilder {
    @PublishedApi
    internal val components: MutableList<WhiskerDefinitionComponent> = mutableListOf()
}

// ----- ModuleDefinition value -----------------------------------------------

/**
 * The assembled definition the framework registers with Lynx at
 * module-init time. Immutable; collected from a
 * [WhiskerModuleDefinitionBuilder] block.
 */
public data class ModuleDefinition(public val components: List<WhiskerDefinitionComponent>) {

    /** Module name (first [WhiskerNameComponent] in the components list). */
    public val name: String?
        get() = components.firstNotNullOfOrNull { (it as? WhiskerNameComponent)?.value }

    /** View block, if any. */
    public val view: WhiskerViewComponent?
        get() = components.firstNotNullOfOrNull { it as? WhiskerViewComponent }

    /** Merged constants from all [WhiskerConstantsComponent] blocks. */
    public val constants: Map<String, Any?>
        get() = buildMap {
            for (c in components) {
                if (c is WhiskerConstantsComponent) putAll(c.values)
            }
        }

    /** Module-level (view-less) [Function] declarations. */
    public val functions: List<WhiskerFunctionComponent>
        get() = components.filterIsInstance<WhiskerFunctionComponent>()

    public companion object {
        /**
         * Builder-style constructor — used as
         * `ModuleDefinition { Name(...); Function(...) { ... } }`.
         */
        public operator fun invoke(
            block: WhiskerModuleDefinitionBuilder.() -> Unit,
        ): ModuleDefinition {
            val b = WhiskerModuleDefinitionBuilder()
            b.block()
            return ModuleDefinition(b.components.toList())
        }
    }
}

// ----- Top-level factories — the DSL surface --------------------------------

// Naming convention: PascalCase, mirroring Expo Modules + the
// Swift side. Kotlin allows top-level fn names of any case.

/** `Name("Foo")` — the module's local tag name. */
public fun WhiskerModuleDefinitionBuilder.Name(value: String): WhiskerDefinitionComponent {
    val c = WhiskerNameComponent(value)
    components.add(c)
    return c
}

/** `Constants("k" to v, ...)` — static dictionary. */
public fun WhiskerModuleDefinitionBuilder.Constants(
    vararg entries: Pair<String, Any?>,
): WhiskerDefinitionComponent {
    val c = WhiskerConstantsComponent(entries.toMap())
    components.add(c)
    return c
}

/** `Constants(mapOf(...))` — same, but takes a Map directly. */
public fun WhiskerModuleDefinitionBuilder.Constants(
    values: Map<String, Any?>,
): WhiskerDefinitionComponent {
    val c = WhiskerConstantsComponent(values)
    components.add(c)
    return c
}

/**
 * `View(MyView::class.java) { ... }` — registers a Lynx UI
 * subclass + its inner DSL block (Prop / Function / Events).
 */
public fun WhiskerModuleDefinitionBuilder.View(
    viewClass: Class<*>,
    block: WhiskerViewDefinitionBuilder.() -> Unit,
): WhiskerDefinitionComponent {
    val b = WhiskerViewDefinitionBuilder()
    b.block()
    val c = WhiskerViewComponent(viewClass, b.components.toList())
    components.add(c)
    return c
}

/** `Events("a", "b", ...)` — variadic event-name declaration. */
public fun WhiskerModuleDefinitionBuilder.Events(
    vararg names: String,
): WhiskerDefinitionComponent {
    val c = WhiskerEventsComponent(names.toList())
    components.add(c)
    return c
}

/** Events declared inside a `View(...)` block. */
public fun WhiskerViewDefinitionBuilder.Events(
    vararg names: String,
): WhiskerDefinitionComponent {
    val c = WhiskerEventsComponent(names.toList())
    components.add(c)
    return c
}

// ----- Prop factories ------------------------------------------------------

/**
 * `Prop("foo") { view, value -> ... }` — view-bearing prop setter.
 * Inside a `View(...)` block.
 *
 * Reified type parameters let the framework downcast at dispatch
 * time. Mismatches silently no-op (with a debug-build log).
 */
public inline fun <reified V : Any, reified T> WhiskerViewDefinitionBuilder.Prop(
    name: String,
    noinline setter: (V, T) -> Unit,
): WhiskerDefinitionComponent {
    val c = WhiskerPropComponent(name) { viewAny: Any, valueAny: Any? ->
        val v = viewAny as? V
        if (v == null) {
            // Type mismatch — drop the call.
            return@WhiskerPropComponent
        }
        @Suppress("UNCHECKED_CAST")
        val t = valueAny as? T
        if (t == null && valueAny != null) {
            return@WhiskerPropComponent
        }
        @Suppress("UNCHECKED_CAST")
        setter(v, t as T)
    }
    components.add(c)
    return c
}

// ----- Function factories — overloads for 0..2 args -------------------------

// View-bound functions: first arg is the view type.

public inline fun <reified V : Any> WhiskerViewDefinitionBuilder.Function(
    name: String,
    crossinline handler: (V) -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { viewAny: Any?, _: List<Any?> ->
        val v = viewAny as? V ?: return@WhiskerFunctionComponent null
        handler(v)
    }
    components.add(c)
    return c
}

public inline fun <reified V : Any, reified A> WhiskerViewDefinitionBuilder.Function(
    name: String,
    crossinline handler: (V, A) -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { viewAny: Any?, args: List<Any?> ->
        val v = viewAny as? V ?: return@WhiskerFunctionComponent null
        val a = args.firstOrNull() as? A ?: return@WhiskerFunctionComponent null
        handler(v, a)
    }
    components.add(c)
    return c
}

public inline fun <reified V : Any, reified A, reified B> WhiskerViewDefinitionBuilder.Function(
    name: String,
    crossinline handler: (V, A, B) -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { viewAny: Any?, args: List<Any?> ->
        val v = viewAny as? V ?: return@WhiskerFunctionComponent null
        if (args.size < 2) return@WhiskerFunctionComponent null
        val a = args[0] as? A ?: return@WhiskerFunctionComponent null
        val b = args[1] as? B ?: return@WhiskerFunctionComponent null
        handler(v, a, b)
    }
    components.add(c)
    return c
}

// Module-level (view-less) functions: no view argument.

public fun WhiskerModuleDefinitionBuilder.Function(
    name: String,
    handler: () -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { _: Any?, _: List<Any?> -> handler() }
    components.add(c)
    return c
}

public inline fun <reified A> WhiskerModuleDefinitionBuilder.Function(
    name: String,
    crossinline handler: (A) -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { _: Any?, args: List<Any?> ->
        val a = args.firstOrNull() as? A ?: return@WhiskerFunctionComponent null
        handler(a)
    }
    components.add(c)
    return c
}

public inline fun <reified A, reified B> WhiskerModuleDefinitionBuilder.Function(
    name: String,
    crossinline handler: (A, B) -> Any?,
): WhiskerDefinitionComponent {
    val c = WhiskerFunctionComponent(name) { _: Any?, args: List<Any?> ->
        if (args.size < 2) return@WhiskerFunctionComponent null
        val a = args[0] as? A ?: return@WhiskerFunctionComponent null
        val b = args[1] as? B ?: return@WhiskerFunctionComponent null
        handler(a, b)
    }
    components.add(c)
    return c
}

File: platforms/ios/Sources/WhiskerModuleApi/ModuleDefinition.swift
// Phase L-2a — `ModuleDefinition` DSL surface (iOS).
//
// Single-language DSL that supersedes the `@WhiskerComponent` /
// `@WhiskerProp` / `@WhiskerUIMethod` annotation set. Modeled after
// Expo Modules' `ModuleDefinition` (https://docs.expo.dev/modules/module-api/)
// but emits direct registrations against Lynx's prop / method
// dispatch tables instead of routing through `@LynxProp` /
// `@LynxUIMethod` reflection.
//
// ## Target syntax
//
// ```swift
// @WhiskerModule
// public final class VideoModule: Module {
//     public override func definition() -> ModuleDefinition {
//         Name("Video")
//
//         Constants(["maxResolution": "1080p"])
//
//         View(WhiskerVideoView.self) {
//             Prop("src") { (view: WhiskerVideoView, value: String) in
//                 view.setSrc(value)
//             }
//             Function("play")  { (view: WhiskerVideoView) in view.play()  }
//             Function("pause") { (view: WhiskerVideoView) in view.pause() }
//             Function("seek")  { (view: WhiskerVideoView, seconds: Double) in
//                 view.seek(seconds)
//             }
//             Events("onCompleted")
//         }
//     }
// }
// ```
//
// View-less modules (function-only) work the same way without the
// inner `View(...)` block:
//
// ```swift
// @WhiskerModule
// public final class LocalStoreModule: Module {
//     public override func definition() -> ModuleDefinition {
//         Name("WhiskerLocalStore")
//         Function("save") { (key: String, value: String) -> Bool in
//             UserDefaults.standard.set(value, forKey: key)
//             return true
//         }
//         Function("load") { (key: String) -> String? in
//             UserDefaults.standard.string(forKey: key)
//         }
//     }
// }
// ```
//
// ## What L-2a delivers
//
// This file defines the **DSL surface and value model**. The
// `Module` base class collects the `ModuleDefinition` at init time;
// the iOS dispatch glue (L-2b) + the `WhiskerComponentsCodegen`
// plugin's `@WhiskerModule` discovery wire it into Lynx's prop /
// method dispatch tables.

import Foundation

// MARK: - Component model

/// Type-erased component the result builder collects. Every DSL
/// factory function returns one of these. Concrete variants live
/// below.
public protocol WhiskerDefinitionComponent {}

// MARK: - Top-level components

/// Module-name component. Exactly one expected per module.
///
/// `Name("Video")` registers the module under the local tag string
/// `Video`; the Whisker build layer prepends the package's cargo
/// crate name to produce the fully-qualified tag
/// (`<crate>:Video`) so two crates can both export a `Video`
/// element without colliding.
public struct WhiskerNameComponent: WhiskerDefinitionComponent {
    public let value: String
    public init(_ value: String) { self.value = value }
}

/// Static constants component. Authors emit a dictionary that
/// the framework exposes to the host. Mirrors Expo's deprecated
/// `Constants([...])` form — we ship the dictionary form only;
/// the dynamic closure form (`Constants { [...] }`) and per-key
/// `Constant("key") { ... }` lazy form can land later.
public struct WhiskerConstantsComponent: WhiskerDefinitionComponent {
    public let values: [String: Any]
    public init(_ values: [String: Any]) { self.values = values }
}

/// View block — registers a Lynx UI subclass + its `Prop` /
/// `Function` / `Events` sub-components. Exactly one expected for
/// view-bearing modules; absent entirely for function-only modules.
public struct WhiskerViewComponent: WhiskerDefinitionComponent {
    /// Type-erased to AnyClass so the call site doesn't need to
    /// generic the parent struct. The concrete class is the Lynx
    /// UI subclass (typically a `LynxUI<UIView>` subclass).
    public let viewClass: AnyClass
    public let components: [WhiskerDefinitionComponent]
    public init(viewClass: AnyClass, components: [WhiskerDefinitionComponent]) {
        self.viewClass = viewClass
        self.components = components
    }
}

// MARK: - View-block components (also legal at module-level for function-only modules)

/// Type-erased setter the framework calls on prop dispatch.
/// `view` is the Lynx UI instance the value was set on; `value`
/// is the raw Lynx value (Whisker value-decoded by L-2b/c).
public typealias WhiskerPropSetter = (_ view: AnyObject, _ value: Any?) -> Void

/// Prop component — a single named setter on the view class.
public struct WhiskerPropComponent: WhiskerDefinitionComponent {
    public let name: String
    public let setter: WhiskerPropSetter
    public init(name: String, setter: @escaping WhiskerPropSetter) {
        self.name = name
        self.setter = setter
    }
}

/// Type-erased function handler. `args` carry through positional
/// arguments from the JS / Rust call site; the result is either
/// the function's return value (auto-encoded) or `nil` for
/// `Void`-returning bodies. `view` is `nil` for module-level
/// `Function`s, the Lynx UI instance for view-block `Function`s.
public typealias WhiskerFunctionHandler = (_ view: AnyObject?, _ args: [Any?]) -> Any?

/// Sync function component. Same shape inside a `View(...)`
/// block (gets the view as `view`) and at module-level
/// (view-less; `view` is `nil`).
public struct WhiskerFunctionComponent: WhiskerDefinitionComponent {
    public let name: String
    public let handler: WhiskerFunctionHandler
    public init(name: String, handler: @escaping WhiskerFunctionHandler) {
        self.name = name
        self.handler = handler
    }
}

/// Event-name declaration component. Authors declare event names
/// they intend to emit; the runtime uses the list for
/// type-checking + future docs generation. The dispatch site
/// stays imperative (`WhiskerCustomEvent.dispatch(from:name:params:)`)
/// — `Events("foo")` just records the name.
public struct WhiskerEventsComponent: WhiskerDefinitionComponent {
    public let names: [String]
    public init(_ names: [String]) { self.names = names }
}

// MARK: - Result builders

/// Top-level `definition() -> ModuleDefinition` body builder.
@resultBuilder
public struct WhiskerModuleDefinitionBuilder {
    public static func buildBlock(_ components: WhiskerDefinitionComponent...) -> ModuleDefinition {
        ModuleDefinition(components: components)
    }
    /// Allow a body that ends with a single component (no trailing
    /// expressions) — Swift's variadic packing covers it but the
    /// explicit overload makes the diagnostic clearer when there's
    /// only one entry.
    public static func buildBlock(_ component: WhiskerDefinitionComponent) -> ModuleDefinition {
        ModuleDefinition(components: [component])
    }
    /// Empty body — yields an empty definition. Mostly used by the
    /// `WhiskerModule.definition()` default impl below.
    public static func buildBlock() -> ModuleDefinition {
        ModuleDefinition(components: [])
    }

    /// Optional / conditional emission.
    public static func buildOptional(
        _ component: WhiskerDefinitionComponent?
    ) -> WhiskerDefinitionComponent {
        component ?? EmptyComponent()
    }
    public static func buildEither(
        first: WhiskerDefinitionComponent
    ) -> WhiskerDefinitionComponent { first }
    public static func buildEither(
        second: WhiskerDefinitionComponent
    ) -> WhiskerDefinitionComponent { second }
}

/// Nested `View(...) { ... }` body builder.
@resultBuilder
public struct WhiskerViewDefinitionBuilder {
    public static func buildBlock(
        _ components: WhiskerDefinitionComponent...
    ) -> [WhiskerDefinitionComponent] {
        components
    }
    public static func buildBlock(
        _ component: WhiskerDefinitionComponent
    ) -> [WhiskerDefinitionComponent] {
        [component]
    }
    public static func buildBlock() -> [WhiskerDefinitionComponent] { [] }

    public static func buildOptional(
        _ component: WhiskerDefinitionComponent?
    ) -> WhiskerDefinitionComponent {
        component ?? EmptyComponent()
    }
    public static func buildEither(
        first: WhiskerDefinitionComponent
    ) -> WhiskerDefinitionComponent { first }
    public static func buildEither(
        second: WhiskerDefinitionComponent
    ) -> WhiskerDefinitionComponent { second }
}

/// Empty / placeholder component for optional-builder paths.
/// Registration phase filters these out.
public struct EmptyComponent: WhiskerDefinitionComponent {
    public init() {}
}

// MARK: - ModuleDefinition value

/// The assembled definition the framework registers with Lynx at
/// module-init time. Immutable; collected from the
/// `@WhiskerModuleDefinitionBuilder` body of `definition()`.
public struct ModuleDefinition {
    public let components: [WhiskerDefinitionComponent]

    public init(components: [WhiskerDefinitionComponent]) {
        self.components = components
    }

    public init(@WhiskerModuleDefinitionBuilder _ body: () -> ModuleDefinition) {
        self = body()
    }

    /// Module name. Returns `nil` if no `Name(...)` was declared —
    /// Phase L-2b will require this and surface a clear error at
    /// registration time.
    public var name: String? {
        for c in components {
            if let n = c as? WhiskerNameComponent { return n.value }
        }
        return nil
    }

    /// View block, if any.
    public var view: WhiskerViewComponent? {
        for c in components {
            if let v = c as? WhiskerViewComponent { return v }
        }
        return nil
    }

    /// Constants dictionary merged from all `Constants(...)` blocks.
    public var constants: [String: Any] {
        var merged: [String: Any] = [:]
        for c in components {
            if let cc = c as? WhiskerConstantsComponent {
                for (k, v) in cc.values { merged[k] = v }
            }
        }
        return merged
    }

    /// Module-level (view-less) functions — i.e. `Function(...)`
    /// declared OUTSIDE a `View(...)` block.
    public var functions: [WhiskerFunctionComponent] {
        components.compactMap { $0 as? WhiskerFunctionComponent }
    }
}

// MARK: - Top-level factories — the DSL surface authors call

// Naming convention: PascalCase to mirror Expo Modules. Swift
// allows top-level function names of any case.

/// `Name("Foo")` — the module's local tag name.
public func Name(_ value: String) -> WhiskerDefinitionComponent {
    WhiskerNameComponent(value)
}

/// `Constants([key: value, ...])` — static constants exposed to
/// the host. Dictionary form only in v1; per-key lazy form
/// (`Constant("k") { ... }`) lands later.
public func Constants(_ values: [String: Any]) -> WhiskerDefinitionComponent {
    WhiskerConstantsComponent(values)
}

/// `View(MyView.self) { ... }` — registers a Lynx UI subclass +
/// its inner DSL block (Prop / Function / Events).
public func View<V: AnyObject>(
    _ viewClass: V.Type,
    @WhiskerViewDefinitionBuilder _ body: () -> [WhiskerDefinitionComponent]
) -> WhiskerDefinitionComponent {
    WhiskerViewComponent(viewClass: viewClass, components: body())
}

/// `Events("a", "b", ...)` — declare event names this module
/// emits. Just metadata; dispatch stays imperative via
/// `WhiskerCustomEvent.dispatch(...)`.
public func Events(_ names: String...) -> WhiskerDefinitionComponent {
    WhiskerEventsComponent(names)
}

// MARK: - Prop factories

/// `Prop("foo") { (view, value) in ... }` — view-bearing prop
/// setter. The closure's parameter types are checked at the call
/// site; the framework runs a downcast at dispatch time and
/// silently no-ops on a mismatch (with a debug-build log).
public func Prop<V: AnyObject, T>(
    _ name: String,
    _ setter: @escaping (V, T) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerPropComponent(name: name) { uiAny, valueAny in
        guard let ui = uiAny as? V else {
            #if DEBUG
            print("WhiskerModuleApi: Prop(\"\(name)\") view type mismatch — expected \(V.self), got \(type(of: uiAny))")
            #endif
            return
        }
        guard let value = valueAny as? T else {
            #if DEBUG
            print("WhiskerModuleApi: Prop(\"\(name)\") value type mismatch — expected \(T.self), got \(type(of: valueAny as Any))")
            #endif
            return
        }
        setter(ui, value)
    }
}

// MARK: - Function factories — 0-arg through 4-arg overloads

// L-2a ships sync `Function` only. `AsyncFunction` lands in L-2d.

/// `Function("noargs") { (view) in ... }` — sync view-bound
/// 0-arg function.
public func Function<V: AnyObject>(
    _ name: String,
    _ handler: @escaping (V) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, _ in
        guard let view = viewAny as? V else { return nil }
        handler(view)
        return nil
    }
}

public func Function<V: AnyObject, R>(
    _ name: String,
    _ handler: @escaping (V) -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, _ in
        guard let view = viewAny as? V else { return nil }
        return handler(view)
    }
}

public func Function<V: AnyObject, A>(
    _ name: String,
    _ handler: @escaping (V, A) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, args in
        guard let view = viewAny as? V, let a = args.first as? A else { return nil }
        handler(view, a)
        return nil
    }
}

public func Function<V: AnyObject, A, R>(
    _ name: String,
    _ handler: @escaping (V, A) -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, args in
        guard let view = viewAny as? V, let a = args.first as? A else { return nil }
        return handler(view, a)
    }
}

public func Function<V: AnyObject, A, B>(
    _ name: String,
    _ handler: @escaping (V, A, B) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, args in
        guard let view = viewAny as? V,
              args.count >= 2,
              let a = args[0] as? A, let b = args[1] as? B
        else { return nil }
        handler(view, a, b)
        return nil
    }
}

public func Function<V: AnyObject, A, B, R>(
    _ name: String,
    _ handler: @escaping (V, A, B) -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { viewAny, args in
        guard let view = viewAny as? V,
              args.count >= 2,
              let a = args[0] as? A, let b = args[1] as? B
        else { return nil }
        return handler(view, a, b)
    }
}

// Module-level (view-less) `Function` — same shape but no view
// arg. Reachable inside a function-only module.

public func Function(
    _ name: String,
    _ handler: @escaping () -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, _ in
        handler()
        return nil
    }
}

public func Function<R>(
    _ name: String,
    _ handler: @escaping () -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, _ in
        handler()
    }
}

public func Function<A>(
    _ name: String,
    _ handler: @escaping (A) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, args in
        guard let a = args.first as? A else { return nil }
        handler(a)
        return nil
    }
}

public func Function<A, R>(
    _ name: String,
    _ handler: @escaping (A) -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, args in
        guard let a = args.first as? A else { return nil }
        return handler(a)
    }
}

public func Function<A, B>(
    _ name: String,
    _ handler: @escaping (A, B) -> Void
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, args in
        guard args.count >= 2, let a = args[0] as? A, let b = args[1] as? B else { return nil }
        handler(a, b)
        return nil
    }
}

public func Function<A, B, R>(
    _ name: String,
    _ handler: @escaping (A, B) -> R
) -> WhiskerDefinitionComponent {
    WhiskerFunctionComponent(name: name) { _, args in
        guard args.count >= 2, let a = args[0] as? A, let b = args[1] as? B else { return nil }
        return handler(a, b)
    }
}
